LibOS/shim/src/shim_init.c

```diff
@@ -442,14 +442,17 @@ int init_manifest (PAL_HANDLE manifest_handle)
 {
     int ret = 0;
     void * addr;
-    size_t size;
-    bool stream_unmap_on_fail = false;
-    bool bkeep_unmap_on_fail = false;
-    struct config_store* new_root_config = NULL;
+    size_t size, map_size;
+
+#define MAP_FLAGS (MAP_PRIVATE|MAP_ANONYMOUS|VMA_INTERNAL)
 
     if (PAL_CB(manifest_preload.start)) {
         addr = PAL_CB(manifest_preload.start);
         size = PAL_CB(manifest_preload.end) - PAL_CB(manifest_preload.start);
+        map_size = ALIGN_UP(size);
+        if (bkeep_mmap(addr, map_size, PROT_READ, MAP_FLAGS,
+                       NULL, 0, "manifest") < 0)
```

Comment:
Shouldn't this bookkeeping call be executed right after the area [manifest_preload.start, manifest_preload.end) was allocated? Memory allocations in between can damage this area.

Reply:
This probably should be moved to the beginning of init_vma(). This corner case means PAL has mapped the manifest file before SHIM starts.